### ISSUE TITLE
[1.16] Performance problem on player join

### ISSUE THREAD

**Issue by DarkArc** [PaperMC/Paper#3796]
###  Timings or Profile link:
https://spark.lucko.me/#ZN4qKDjCbt

### Description of issue:
There seems to be an issue with `net.minecraft.server.v1_16_R1.AdvancementDataPlayer.e()`. It almost looks like infinite recursion from the sampling, but I don't see the StackOverflowError I would expect.

### Paper build number:
Equivalent to: https://github.com/PaperMC/Paper/commit/4fd17445297b41e7121303bf886b32192346e715

**Comment by DarkArc** [comment]
I made a little patch to record the traversal:
```
diff --git a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
index 57b9d1344..99be7aaed 100644
--- a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+++ b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
@@ -15,12 +15,11 @@ import com.google.gson.stream.JsonReader;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.JsonOps;
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.IOException;
-import java.io.OutputStreamWriter;
-import java.io.StringReader;
+
+import java.io.*;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
 import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
@@ -438,6 +437,25 @@ public class AdvancementDataPlayer {
         this.data.put(advancement, advancementprogress);
     }
 
+    private int depth = 0;
+
+    private void eTracked(Advancement advancement) {
+        ++depth;
+
+        try (BufferedWriter writer = java.nio.file.Files.newBufferedWriter(
+            Paths.get("output.txt"), StandardOpenOption.CREATE, StandardOpenOption.APPEND)) {
+            for (int i = depth; i > 0; --i) {
+                writer.write(" ");
+            }
+            writer.write(advancement.getName() + "\n");
+        } catch (IOException ex) {
+            ex.printStackTrace();
+        }
+
+        this.e(advancement);
+        --depth;
+    }
+
     private void e(Advancement advancement) {
         boolean flag = this.f(advancement);
         boolean flag1 = this.h.contains(advancement);
@@ -454,7 +472,7 @@ public class AdvancementDataPlayer {
         }
 
         if (flag != flag1 && advancement.b() != null) {
-            this.e(advancement.b());
+            eTracked(advancement.b());
         }
 
         Iterator iterator = advancement.e().iterator();
@@ -462,7 +480,7 @@ public class AdvancementDataPlayer {
         while (iterator.hasNext()) {
             Advancement advancement1 = (Advancement) iterator.next();
 
-            this.e(advancement1);
+            eTracked(advancement1);
         }
 
     }

```

This algorithm is running `O(N^2)`, I'll spare the full ouput but here's the beginning:
```
 minecraft:recipes/root
  minecraft:recipes/building_blocks/jungle_slab
   minecraft:recipes/root
    minecraft:recipes/building_blocks/jungle_slab
    minecraft:recipes/misc/quartz_from_blasting
     minecraft:recipes/root
      minecraft:recipes/building_blocks/jungle_slab
      minecraft:recipes/misc/quartz_from_blasting
      minecraft:recipes/building_blocks/polished_granite_slab
       minecraft:recipes/root
        minecraft:recipes/building_blocks/jungle_slab
        minecraft:recipes/misc/quartz_from_blasting
        minecraft:recipes/building_blocks/polished_granite_slab
        minecraft:recipes/building_blocks/acacia_planks
         minecraft:recipes/root
          minecraft:recipes/building_blocks/jungle_slab
          minecraft:recipes/misc/quartz_from_blasting
          minecraft:recipes/building_blocks/polished_granite_slab
          minecraft:recipes/building_blocks/acacia_planks
          minecraft:recipes/redstone/repeater
           minecraft:recipes/root
            minecraft:recipes/building_blocks/jungle_slab
            minecraft:recipes/misc/quartz_from_blasting
            minecraft:recipes/building_blocks/polished_granite_slab
            minecraft:recipes/building_blocks/acacia_planks
            minecraft:recipes/redstone/repeater
            minecraft:recipes/tools/compass
             minecraft:recipes/root
              minecraft:recipes/building_blocks/jungle_slab
              minecraft:recipes/misc/quartz_from_blasting
              minecraft:recipes/building_blocks/polished_granite_slab
              minecraft:recipes/building_blocks/acacia_planks
              minecraft:recipes/redstone/repeater
              minecraft:recipes/tools/compass
              minecraft:recipes/redstone/oak_button
               minecraft:recipes/root
                minecraft:recipes/building_blocks/jungle_slab
                minecraft:recipes/misc/quartz_from_blasting
                minecraft:recipes/building_blocks/polished_granite_slab
                minecraft:recipes/building_blocks/acacia_planks
                minecraft:recipes/redstone/repeater
                minecraft:recipes/tools/compass
                minecraft:recipes/redstone/oak_button
                minecraft:recipes/building_blocks/quartz_block
                 minecraft:recipes/root
                  minecraft:recipes/building_blocks/jungle_slab
                  minecraft:recipes/misc/quartz_from_blasting
                  minecraft:recipes/building_blocks/polished_granite_slab
                  minecraft:recipes/building_blocks/acacia_planks
                  minecraft:recipes/redstone/repeater
                  minecraft:recipes/tools/compass
                  minecraft:recipes/redstone/oak_button
                  minecraft:recipes/building_blocks/quartz_block
                  minecraft:recipes/redstone/dark_oak_pressure_plate
                   minecraft:recipes/root
                    minecraft:recipes/building_blocks/jungle_slab
                    minecraft:recipes/misc/quartz_from_blasting
                    minecraft:recipes/building_blocks/polished_granite_slab
                    minecraft:recipes/building_blocks/acacia_planks
                    minecraft:recipes/redstone/repeater
                    minecraft:recipes/tools/compass
                    minecraft:recipes/redstone/oak_button
                    minecraft:recipes/building_blocks/quartz_block
                    minecraft:recipes/redstone/dark_oak_pressure_plate
                    minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                     minecraft:recipes/root
                      minecraft:recipes/building_blocks/jungle_slab
                      minecraft:recipes/misc/quartz_from_blasting
                      minecraft:recipes/building_blocks/polished_granite_slab
                      minecraft:recipes/building_blocks/acacia_planks
                      minecraft:recipes/redstone/repeater
                      minecraft:recipes/tools/compass
                      minecraft:recipes/redstone/oak_button
                      minecraft:recipes/building_blocks/quartz_block
                      minecraft:recipes/redstone/dark_oak_pressure_plate
                      minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                      minecraft:recipes/food/cooked_beef
                       minecraft:recipes/root
                        minecraft:recipes/building_blocks/jungle_slab
                        minecraft:recipes/misc/quartz_from_blasting
                        minecraft:recipes/building_blocks/polished_granite_slab
                        minecraft:recipes/building_blocks/acacia_planks
                        minecraft:recipes/redstone/repeater
                        minecraft:recipes/tools/compass
                        minecraft:recipes/redstone/oak_button
                        minecraft:recipes/building_blocks/quartz_block
                        minecraft:recipes/redstone/dark_oak_pressure_plate
                        minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                        minecraft:recipes/food/cooked_beef
                        minecraft:recipes/building_blocks/smooth_quartz_stairs_from_smooth_quartz_stonecutting
                         minecraft:recipes/root
                          minecraft:recipes/building_blocks/jungle_slab
                          minecraft:recipes/misc/quartz_from_blasting
                          minecraft:recipes/building_blocks/polished_granite_slab
                          minecraft:recipes/building_blocks/acacia_planks
                          minecraft:recipes/redstone/repeater
                          minecraft:recipes/tools/compass
                          minecraft:recipes/redstone/oak_button
                          minecraft:recipes/building_blocks/quartz_block
                          minecraft:recipes/redstone/dark_oak_pressure_plate
                          minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                          minecraft:recipes/food/cooked_beef
                          minecraft:recipes/building_blocks/smooth_quartz_stairs_from_smooth_quartz_stonecutting
                          minecraft:recipes/redstone/iron_door
                           minecraft:recipes/root
                            minecraft:recipes/building_blocks/jungle_slab
                            minecraft:recipes/misc/quartz_from_blasting
                            minecraft:recipes/building_blocks/polished_granite_slab
                            minecraft:recipes/building_blocks/acacia_planks
                            minecraft:recipes/redstone/repeater
                            minecraft:recipes/tools/compass
                            minecraft:recipes/redstone/oak_button
                            minecraft:recipes/building_blocks/quartz_block
                            minecraft:recipes/redstone/dark_oak_pressure_plate
                            minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                            minecraft:recipes/food/cooked_beef
                            minecraft:recipes/building_blocks/smooth_quartz_stairs_from_smooth_quartz_stonecutting
                            minecraft:recipes/redstone/iron_door
                            minecraft:recipes/decorations/nether_brick_fence
                             minecraft:recipes/root
                              minecraft:recipes/building_blocks/jungle_slab
                              minecraft:recipes/misc/quartz_from_blasting
                              minecraft:recipes/building_blocks/polished_granite_slab
                              minecraft:recipes/building_blocks/acacia_planks
                              minecraft:recipes/redstone/repeater
                              minecraft:recipes/tools/compass
                              minecraft:recipes/redstone/oak_button
                              minecraft:recipes/building_blocks/quartz_block
                              minecraft:recipes/redstone/dark_oak_pressure_plate
                              minecraft:recipes/building_blocks/mossy_stone_brick_slab_from_mossy_stone_brick_stonecutting
                              minecraft:recipes/food/cooked_beef
                              minecraft:recipes/building_blocks/smooth_quartz_stairs_from_smooth_quartz_stonecutting
                              minecraft:recipes/redstone/iron_door
                              minecraft:recipes/decorations/nether_brick_fence
                              minecraft:recipes/building_blocks/chiseled_polished_blackstone_from_blackstone_stonecutting
                               minecraft:recipes/root
```

**Comment by electronicboy** [comment]
That makes a lot of sense based upon what I’ve seen of crash reports, etc, but; Just…

Oh god

**Comment by DarkArc** [comment]
Fixed in https://github.com/PaperMC/Paper/commit/97cedaa5891f31028ed493d160773ce4d32220d9

**Comment by aikar** [comment]
@DarkArc i was questioning, when it goes to the parent, it then checks that nodes parent again, shouldnt we keep the PARENT_OF_ITERATOR there? as if it does go up another level, itll now be marked as ROOT again and be back in same boat

**Comment by DarkArc** [comment]
You could move this check into the parent and say "if I'm from a child iterator don't event look at my parent", you could also move it, and say "if I'm a parent of a child, ignore the flags". I figured if the flags had actually changed though, and the vanilla algorithm would iterate again, we should iterate again.

That decision was largely based on not fully understanding what this function is trying to do. I can say that in my version of this that prints the output, I'm fairly confident that doesn't happen, and I'm almost certain that even if it does, it's on a small subset, rather than the primary "list" of advancements. My printing code went from generating a file that was a few dozen megabytes from a single login, to a few hundred kilobytes if that.

Having the unobfuscated code would probably be helpful here (to try and really understand the original intent)... I know mojang was releasing source mappings; I'd have to look into how to actually get those and use them on 1.16 to get a source file. If you know of any good resources on how to do that... that would be much appreciated, and I could probably give a better answer.

----

Basically this is an attempt to cull (what appears to be) unintentional iteration, without culling (what might be) intentional iteration.
